common/persistence/data_store_interfaces_test.go

```diff
@@ -0,0 +1,177 @@
+		}, "only thriftrw data can start with Y without panicking")
+		// all others panic
+		for _, encoding := range allEncodings {
+			if encoding == common.EncodingTypeThriftRW {
+				continue // handled above
+			}
+			assert.Panicsf(t, func() {
+				NewDataBlob([]byte(problematic), encoding)
+			}, "non-thriftrw %v should panic if first byte is Y", encoding)
+		}
+
+		// make sure other data/encoding combinations do not panic for any encoding
+		for _, dat := range []string{"Z...", "other"} {
+			for _, encoding := range allEncodings {
+				assert.NotPanicsf(t, func() {
+					NewDataBlob([]byte(dat), encoding)
+				}, "should not panic with %q on encoding %q", dat, encoding)
+			}
+		}
+	})
+	t.Run("from", func(t *testing.T) {
+		assertEmpty := func(b []byte, s string) {
+			assert.Nil(t, b, "should have nil bytes")
+			assert.Empty(t, s, "should have empty string")
+		}
+		assertEmpty(FromDataBlob(nil))
+		assertEmpty(FromDataBlob(&DataBlob{}))
+		assertEmpty(FromDataBlob(&DataBlob{Encoding: common.EncodingTypeThriftRW}))
+
+		dat, str := FromDataBlob(&DataBlob{
+			Data:     []byte("data"),
+			Encoding: common.EncodingTypeJSON,
+		})
+		assert.Equal(t, []byte("data"), dat, "data should be returned")
+		assert.Equal(t, string(common.EncodingTypeJSON), str, "encoding should be returned as a string")
+	})
+	t.Run("ToNilSafeDataBlob", func(t *testing.T) {
+		orig := &DataBlob{Encoding: common.EncodingTypeGob} // anything not empty
+		assert.Equal(t, orig, orig.ToNilSafeDataBlob())
+		assert.NotNil(t, (*DataBlob)(nil).ToNilSafeDataBlob(), "typed nils should convert to non-nils")
+	})
+	t.Run("GetEncodingString", func(t *testing.T) {
+		assert.Equal(t, "", (*DataBlob)(nil).GetEncodingString())
+		assert.Equal(t, "test", (&DataBlob{Encoding: "test"}).GetEncodingString())
+	})
+	t.Run("GetData", func(t *testing.T) {
+		// this method returns empty slices, not nils.
+		// I'm not sure if this needs to be maintained, but it must be checked before changing.
+		assert.Equal(t, []byte{}, (*DataBlob)(nil).GetData())
+		assert.Equal(t, []byte{}, (&DataBlob{}).GetData())
+		assert.Equal(t, []byte{}, (&DataBlob{Data: []byte{}}).GetData())
+		assert.Equal(t, []byte("test"), (&DataBlob{Data: []byte("test")}).GetData())
+	})
+	t.Run("GetEncoding", func(t *testing.T) {
+		same := func(encoding common.EncodingType) {
+			assert.Equal(t, encoding, (&DataBlob{Encoding: encoding}).GetEncoding())
+		}
+		unknown := func(encoding common.EncodingType) {
+			assert.Equal(t, common.EncodingTypeUnknown, (&DataBlob{Encoding: encoding}).GetEncoding())
+		}
+		// obvious
+		same(common.EncodingTypeGob)
+		same(common.EncodingTypeJSON)
+		same(common.EncodingTypeThriftRW)
+		same(common.EncodingTypeEmpty)
+
+		// highly suspicious
+		unknown(common.EncodingTypeProto)
+
+		// should be unknown
+		unknown(common.EncodingTypeUnknown)
+		unknown("any other value")
+	})
+	t.Run("To and FromInternal", func(t *testing.T) {
```

Comment:
@Groxx  What do you think about doing some round-trip testing against the various encoding methods - if they're not there already (I haven't looked at the existing coverage). 

For all bidirectional encodings, we should be able to round-trip any generated that:
```
for each encoding type:
    blob := generate(type)
    roundtripped := blob.FromInternal(blob.ToInternal())
    assert.Equal(t, blob, roundtripped)
```

Reply:
means to create data is possible with `testing/quick` for limited stuff and `common/testing/testdatagen/idlfuzzedtestdata` might be useful if you need either inspiration or concrete types

Reply:
if we're covering this already, feel free to ignore

Reply:
Yeah, an explicit "round trips with X" seems semantically-meaningful, as it's not just "they look similar".  I'll add that 👍